### DISCUSSION TITLE
Is it possible to save RK steps simulation?

### DISCUSSION THREAD

**Discussion by Marcodelloro** [casadi/casadi#3475] [Q&A]
Hi everybody, 
I'm currently working on my MSc Thesis and I am using casadi to solve some optimization problems. 
I have a casadi function f = @(X, coefs) depeding on X _(8x399 opti.variable)_ and coefs _(13x399 opti.variable)_.

When computing RK4, it takes a lot of time to set all the  _opti.subject_to_ constaints, at least I guess that is the code part taking so long.

for k = 1:N-1 

% Runge-Kutta 4 integration

   k1 = f(X(:,k),         coefs(:,k));
   k2 = f(X(:,k)+dt/2*k1, coefs(:,k));
   k3 = f(X(:,k)+dt/2*k2, coefs(:,k));
   k4 = f(X(:,k)+dt*k3,   coefs(:,k));
   x_next = X(:,k) + dt/6*(k1+2*k2+2*k3+k4);
   **full_x(:,k) = x_next;**
   opti.subject_to(X(:,k+1) == x_next); % close the gaps

end

Is there any way to make it faster? I wanted to save all x_next steps in _**full_x(:,k) = x_next;**_ , so that i could load it once anf have all the RK timesteps already computed, to just have a for loop compiling the _opti.subject_to_ constaints, but I do not know how to save such MX file (I tried, it gives some "Serializing SWIG objects not supported" error).
Can somebody help me out ?

P.S. I'm very bad at programming and i just started out on CasADi, so it might be a super dumb question, sry

**Comment by sandeep026** [comment]
Dear @Marcodelloro , avoid for loops and please use the **map**  functionality of casadi. It saves a lot of time spent in construction. To further speed up the process, you also evaluate the rk45 step in parallel. Hope that helps.

You can find more information in the links below.
https://github.com/casadi/casadi/wiki/FAQ%3A-How-to-use-map()-and-mapaccum()-to-speed-up-calculations%3F
https://github.com/casadi/casadi/blob/f03af19249932cfaa8ba78871fb2cd7cff0057e5/docs/examples/python/sysid.py

**Comment by jgillis** [comment]
How did you initialize `full_x`?
`full_x = MX(m,n)` will be extremely slow.
`full_x = MX.zeros(m,n)` will be okayish.
`full_x = {}` paired with `full_x{end+1} = x_next` will be good.

Saving CasADi expression is not trivial, but saving CasADi Functions is.
